### PR TITLE
Pbi 237 external data support via url

### PR DESCRIPTION
**Pull request by GagsdName**

### CODE REVIEW

File: projects/make-a-vis/src/lib/data-view/table/table.component.scss [hunk 1]
```diff
@@ -80,6 +80,11 @@
             padding: 0.5rem;
             min-width: 12rem;
             max-width: 12rem;
+            max-height: 4.5rem;
```
Comment (bherr2) [inline projects/make-a-vis/src/lib/data-view/table/table.component.scss]: Do we really need max-height now that we're cutting off the string lengths?

File: projects/make-a-vis/src/lib/data-view/table/table.component.ts [hunk 2]
```diff
@@ -1,17 +1,17 @@
...
-import { combineLatest as rxCombineLatest, Subscription } from 'rxjs';
+import { combineLatest as rxCombineLatest, Observable, of, Subscription } from 'rxjs';
 import { distinctUntilChanged as rxDistinctUntilChanged, map as rxMap } from 'rxjs/operators';
 
 import { getOpenGVGroupPanelsSelector, isGVPanelOpenSelector } from '../../mav-selection/shared/store';
 import { ActionDispatcherService } from '../../shared/services/actionDispatcher/action-dispatcher.service';
 import { DataVariableHoverService } from '../../shared/services/hover/data-variable-hover.service';
 import { DataService, DataSource } from '../shared/data.service';
 import { ExportTableService } from '../shared/export-table.service';
+import { logging } from 'protractor';
```
Comment (bherr2) [inline projects/make-a-vis/src/lib/data-view/table/table.component.ts]: This isn't used in the code

File: projects/make-a-vis/src/lib/data-view/table/table.component.ts [hunk 3]
```diff
@@ -144,6 +163,28 @@ export class TableComponent implements AfterViewInit, OnChanges, OnDestroy {
...
+    if (this.paginatorSubscription) {
+      this.paginatorSubscription.unsubscribe();
+    }
+  }
+
+  /**
+  * Updates data
+  */
+  updateData(): void {
+    this.dataSubscription = this.data$.subscribe((data) => {
```
Comment (bherr2) [inline projects/make-a-vis/src/lib/data-view/table/table.component.ts]: Subscription leak... Need to unsubscribe from the previous data subscription (if it exists) before creating a new one.

File: projects/make-a-vis/src/lib/data-view/table/table.component.ts [hunk 4]
```diff
@@ -37,6 +37,11 @@ export class TableComponent implements AfterViewInit, OnChanges, OnDestroy {
    */
   @Input() tableIndex: number;
 
+  /**
+   * Size of cell value (after which cell value will be truncated and replaced with an ellipsis)
+   */
+  @Input() cellValueSize = 64;
```
Comment (bherr2) [inline projects/make-a-vis/src/lib/data-view/table/table.component.ts]: This is not a good name. It doesn't clearly show that its a maximum or that the number is based on string length (is size pixels, inches, string length, etc??).

A more appropriate name would be `maxCellStringLength`

File: projects/make-a-vis/src/lib/data-view/shared/data.service.ts [hunk 5]
```diff
@@ -70,18 +100,106 @@ export class DataService {
       });
   }
 
+  /**
```
Comment (bherr2) [inline projects/make-a-vis/src/lib/data-view/shared/data.service.ts]: This feels like a lot of code to do a simple thing. getDisplayValue, checkSize, getTooltip, could easily be removed with a couple more lines in injectDisplayVariables.

File: projects/make-a-vis/src/lib/data-view/shared/data.service.ts [hunk 6]
```diff
@@ -50,18 +71,27 @@ export class DataService {
...
             const operator = this.getDataMappingOperator(recordSet.dataVariables, project.graphicVariables, recordSet.id);
             if (recordSet.defaultRecordStream) {
-              recordSet.defaultRecordStream.asObservable().subscribe((changeSet: RawChangeSet<any>) => {
-                dataSource.data = (changeSet.insert || []).map(operator.getter);
-                dataSource.numRows += (changeSet.insert.length || 0) - (changeSet.remove.length || 0);
-                if (changeSet.insert.length > 1) {
-                  dataSource.label = recordSet.labelPlural;
-                }
-              });
+              const data: any[] = [];
```
Comment (bherr2) [inline projects/make-a-vis/src/lib/data-view/shared/data.service.ts]: Logic is wrong here. this data array is never updated. Works with static data, but will not work with streaming data. You should be saving the results of this.updateData back into data.

File: projects/make-a-vis/src/lib/data-view/shared/data.service.ts [hunk 7]
```diff
@@ -50,18 +71,27 @@ export class DataService {
...
+              dataSource.data = recordSet.defaultRecordStream.asObservable().pipe(
+                map((changeSet: RawChangeSet<any>) => {
+                  dataSource.numRows += (changeSet.insert.length || 0) - (changeSet.remove.length || 0);
```
Comment (bherr2) [inline projects/make-a-vis/src/lib/data-view/shared/data.service.ts]: Rather then using changeSet, use the cached data to have a more accurate numRows and label

File: projects/make-a-vis/src/lib/data-view/shared/export-table.service.ts [hunk 8]
```diff
@@ -1,28 +1,45 @@
...
    */
-  save(source: DataSource): void {
-    const header = loMap(source.columns, 'label');
-    const ids = loMap(source.columns, 'id');
-    const values = loMap(source.data, item => loAt(item, ids));
-    const csv = unparse({ fields: header, data: values });
-    const blob = new Blob([csv], { type: 'text/csv' });
-    saveAs(blob, source.label + '.csv');
+   save(source: DataSource): void {
+    this.dataSubscription = source.data.subscribe((data) => {
```
Comment (bherr2) [inline projects/make-a-vis/src/lib/data-view/shared/export-table.service.ts]: subscription leak. Check if a dataSubscription is already there and unsubscribe. Or better yet, have the subscriber close it after one emission (takeOne) that way subscriptions unsubscribe automatically.

File: projects/make-a-vis/src/lib/data-view/main/main.component.ts [hunk 9]
```diff
@@ -9,13 +9,23 @@ import { DataService, DataSource } from '../shared/data.service';
...
+  dataSubscription: Subscription;
+  /**
+   * Boolean to show initial info message in the data view
+   */
   showDataViewMessage = true;
 
   constructor(private dataService: DataService) {
-    this.dataSources = dataService.dataSourcesChanged;
-    this.dataSources.subscribe((d) => {
+    this.dataSources = this.dataService.dataSourcesChanged;
```
Comment (bherr2) [inline projects/make-a-vis/src/lib/data-view/main/main.component.ts]: No good reason to add `this.` to dataService.dataSourcesChanged.

Comment (GagsdName) [reply projects/make-a-vis/src/lib/data-view/main/main.component.ts]: VSCode shows the dataservice instance in constructor arguments as greyed out without `this`, says value is declared but never read.

Comment (bherr2) [reply projects/make-a-vis/src/lib/data-view/main/main.component.ts]: Oh I see, you can remove the private keyword to make it a local (to the constructor) variable. That should stop it from complaining.

**Pull request merged**